### PR TITLE
chore: squeeze more simps

### PR DESCRIPTION
**Pull request by adomani**
Another batch of squeezed `simp`s, converted mostly to `simp only`, but also `simpa`, `tauto`, `aesop` and `simp? says`.

Ref: #11246

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Products/Basic.lean [hunk 1]
```diff
@@ -73,7 +73,10 @@ theorem isIso_prod_iff {P Q : C} {S T : D} {f : (P, S) ⟶ (Q, T)} :
   · rintro ⟨⟨g₁, hfg₁, hgf₁⟩, ⟨g₂, hfg₂, hgf₂⟩⟩
     dsimp at hfg₁ hgf₁ hfg₂ hgf₂
     refine' ⟨⟨(g₁, g₂), _, _⟩⟩
-    repeat { simp; constructor; assumption; assumption }
+    all_goals
+      simp only [prod_Hom, prod_comp, prod_id, Prod.mk.injEq]
+      constructor <;>
+      assumption
```
Comment (mcdoll) [inline Mathlib/CategoryTheory/Products/Basic.lean:79]:
```
    exact ⟨⟨(g₁, g₂), by aesop, by aesop⟩⟩
```

File: Mathlib/Algebra/Order/Ring/WithTop.lean [hunk 2]
```diff
@@ -71,12 +71,12 @@ lemma mul_eq_top_iff : a * b = ⊤ ↔ a ≠ 0 ∧ b = ⊤ ∨ a = ⊤ ∧ b ≠
...
 
 lemma mul_coe_eq_bind {b : α} (hb : b ≠ 0) : ∀ a, (a * b : WithTop α) = a.bind fun a ↦ ↑(a * b)
-  | ⊤ => by simp [top_mul, hb]; rfl
+  | ⊤ => by simp only [ne_eq, coe_eq_zero, hb, not_false_eq_true, top_mul]; rfl
   | (a : α) => rfl
 #align with_top.mul_coe WithTop.mul_coe_eq_bind
 
 lemma coe_mul_eq_bind {a : α} (ha : a ≠ 0) : ∀ b, (a * b : WithTop α) = b.bind fun b ↦ ↑(a * b)
-  | ⊤ => by simp [top_mul, ha]; rfl
+  | ⊤ => by simp only [ne_eq, coe_eq_zero, ha, not_false_eq_true, mul_top]; rfl
```
[the remark this answers is not in the document]
Comment (adomani) [reply Mathlib/Algebra/Order/Ring/WithTop.lean:79]: Thanks!

File: Mathlib/Algebra/Order/Ring/WithTop.lean [hunk 3]
```diff
@@ -71,12 +71,12 @@ lemma mul_eq_top_iff : a * b = ⊤ ↔ a ≠ 0 ∧ b = ⊤ ∨ a = ⊤ ∧ b ≠
 #align with_top.mul_eq_top_iff WithTop.mul_eq_top_iff
 
 lemma mul_coe_eq_bind {b : α} (hb : b ≠ 0) : ∀ a, (a * b : WithTop α) = a.bind fun a ↦ ↑(a * b)
-  | ⊤ => by simp [top_mul, hb]; rfl
+  | ⊤ => by simp only [ne_eq, coe_eq_zero, hb, not_false_eq_true, top_mul]; rfl
```
[the remark this answers is not in the document]
Comment (adomani) [reply Mathlib/Algebra/Order/Ring/WithTop.lean:74]: Thanks!

File: Mathlib/CategoryTheory/Products/Basic.lean (hunk 1 above)
Comment (adomani) [reply Mathlib/CategoryTheory/Products/Basic.lean:79]: Thanks!

File: Mathlib/Control/Bitraversable/Instances.lean [hunk 4]
```diff
@@ -46,7 +46,10 @@ instance : Bitraversable Prod where bitraverse := @Prod.bitraverse
 
 instance : LawfulBitraversable Prod := by
   constructor <;> intros <;> casesm _ × _ <;>
-    simp [bitraverse, Prod.bitraverse, functor_norm] <;> rfl
+    simp only [bitraverse, Prod.bitraverse, Function.comp_apply, Functor.Comp.map_mk,
```
Comment (jcommelin) [inline Mathlib/Control/Bitraversable/Instances.lean:49]: Here also, I'm inclined to make an exception for `simp [...] <;> rfl`. The problem with non-terminal simps is that they are hard to maintain. But `simp; rfl` is an exception to the general rule, I think.
And the squeeze is obscuring the proof, I think...

File: Mathlib/Data/Seq/WSeq.lean [hunk 5]
```diff
@@ -1791,9 +1791,16 @@ theorem join_join (SS : WSeq (WSeq (WSeq α))) : join (join SS) ~ʷ join (map jo
...
+      induction' s using WSeq.recOn with a s s <;>
+        simp only [nil_append, cons_append, destruct_cons, destruct_pure, liftRelAux_inl_inl,
+          LiftRelO, true_and, think_append, destruct_think, Computation.destruct_think,
+          liftRelAux_inr_inr]
+      · induction' S using WSeq.recOn with s S S <;>
+          simp only [nil_append, join_nil, cons_append, join_cons, destruct_think,
+            Computation.destruct_think, think_append, append_assoc, liftRelAux_inr_inr, join_think]
+        · induction' SS using WSeq.recOn with S SS SS <;>
+            simp only [join_nil, destruct_nil, destruct_pure, map_nil, liftRelAux_inl_inl, LiftRelO,
+              join_cons, join_think, destruct_think, Computation.destruct_think, map_cons]
```
Comment (joneugster) [inline Mathlib/Data/Seq/WSeq.lean:1803]: Is this one of the instances where avoiding non-terminal `simp` obfuscates the proof unreasonably?

**Review by joneugster — APPROVED** [review]
@adomani do you want to resolve all merge conflicts once more? Then I think we could see that we get this merged quickly.

(and maybe do a bench! once you've done that, just for good measure)

**Comment by joneugster** [comment]
@adomani : I feel like this PR has been lying around for too long. The squeezes here either already have been fixed or should be addressed by your "flexible linter".

I think it's probably easier to close this PR and redo the things again on a fresh master branch if still relevant.

Please reopen the PR if you disagree.

**Pull request closed without merging**